Manejo de DataFrames de pandas.
Lo primero es visualizar los datos mediante varios métodos.

cargaremos unos datos que tienen ciertas particularidades. Se puede leer como un csv, pero es un .data. No tiene encabezado. Es decir, la primera línea ya son datos.
Es un archivo pensado para tareas de clasificación.


In [1]:
import pandas as pd

In [7]:
file_path = "D:/Descargas/Conservar/copia seguridad antigua/Descargas/Fran/importantes/ESTUDIOS/Cursos externos/UDEMY/Manual de Referencia Data Science ML Python/recursos/Seccion4/Datos/DatosAdult/adult.data"

In [10]:
col_names= ['age',
          'workclass',
          'fnlwgt',
          'education', 
          'education-num',
          'marital-status',
          'occupation',
          'relationship',
          'race',
          'sex',
          'capital-gain',
          'capital-loss',
          'hours-per-week',
          'native-country',
          'income']

In [12]:
df = pd.read_csv(filepath_or_buffer=file_path, header=0, names=col_names)
df
# header 0 y names !=None, es lo que permite que no tome la primera fila como el encabezado. names tendrá que ser una lista que le demos para los encabezados
# sep=',' es el parámetro por defecto, por lo que no hace falta que indequemos nada

age          workclass  fnlwgt    education  education-num  \
0       50   Self-emp-not-inc   83311    Bachelors             13   
1       38            Private  215646      HS-grad              9   
2       53            Private  234721         11th              7   
3       28            Private  338409    Bachelors             13   
4       37            Private  284582      Masters             14   
...    ...                ...     ...          ...            ...   
32555   27            Private  257302   Assoc-acdm             12   
32556   40            Private  154374      HS-grad              9   
32557   58            Private  151910      HS-grad              9   
32558   22            Private  201490      HS-grad              9   
32559   52       Self-emp-inc  287927      HS-grad              9   

            marital-status          occupation    relationship    race  \
0       Married-civ-spouse     Exec-managerial         Husband   White   
1                 Divorced   Handlers-cleaners   Not-in-family   White   
2       Married-civ-spouse   Handlers-cleaners         Husband   Black   
3       Married-civ-spouse      Prof-specialty            Wife   Black   
4       Married-civ-spouse     Exec-managerial            Wife   White   
...                    ...                 ...             ...     ...   
32555   Married-civ-spouse        Tech-support            Wife   White   
32556   Married-civ-spouse   Machine-op-inspct         Husband   White   
32557              Widowed        Adm-clerical       Unmarried   White   
32558        Never-married        Adm-clerical       Own-child   White   
32559   Married-civ-spouse     Exec-managerial            Wife   White   

           sex  capital-gain  capital-loss  hours-per-week  native-country  \
0         Male             0             0              13   United-States   
1         Male             0             0              40   United-States   
2         Male             0             0              40   United-States   
3       Female             0             0              40            Cuba   
4       Female             0             0              40   United-States   
...        ...           ...           ...             ...             ...   
32555   Female             0             0              38   United-States   
32556     Male             0             0              40   United-States   
32557   Female             0             0              40   United-States   
32558     Male             0             0              20   United-States   
32559   Female         15024             0              40   United-States   

       income  
0       <=50K  
1       <=50K  
2       <=50K  
3       <=50K  
4       <=50K  
...       ...  
32555   <=50K  
32556    >50K  
32557   <=50K  
32558   <=50K  
32559    >50K  

[32560 rows x 15 columns]

Al cargar los datos nos damos cuenta que los nombres de los encabezados no coinciden con las columnas, porque le hemos dado menos nombres que columnas hay, porque no hemos pasado en los nombres de las columnas la última que coincide con el de la variable objetivo de clasificaicón, si es máyor o menor de 50k. la primera columna de nuestro DF lo toma como índice, cuando en realidad es la edad. Por lo que esto hay que controlarlo y añadir el nombre de la última variable manualmente.
Ahora esta bien, lo he cambiado antes.
Ahora se crea un índice automatico incremental. empezando en 0 y acabando en 32559.

<h3>Visualización de datos con algunas funciones<br>
Suele ser lo primero que se hace cuando cargamos un DF que no conocemos a priori.<br> Es un primer chequeo</h3>

In [13]:
""" método de instancia que permite ver las primeras filas del df. Por defecto muestra 5 primeras filas, pero se puede cambiar, 
ya sea poniendo un número concreto n=nº o un porcentaje sobre el total del df 
"""
df.head()

age          workclass  fnlwgt   education  education-num  \
0   50   Self-emp-not-inc   83311   Bachelors             13   
1   38            Private  215646     HS-grad              9   
2   53            Private  234721        11th              7   
3   28            Private  338409   Bachelors             13   
4   37            Private  284582     Masters             14   

        marital-status          occupation    relationship    race      sex  \
0   Married-civ-spouse     Exec-managerial         Husband   White     Male   
1             Divorced   Handlers-cleaners   Not-in-family   White     Male   
2   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
3   Married-civ-spouse      Prof-specialty            Wife   Black   Female   
4   Married-civ-spouse     Exec-managerial            Wife   White   Female   

   capital-gain  capital-loss  hours-per-week  native-country  income  
0             0             0              13   United-States   <=50K  
1             0             0              40   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40            Cuba   <=50K  
4             0             0              40   United-States   <=50K

In [14]:
df.head(n=3)

age          workclass  fnlwgt   education  education-num  \
0   50   Self-emp-not-inc   83311   Bachelors             13   
1   38            Private  215646     HS-grad              9   
2   53            Private  234721        11th              7   

        marital-status          occupation    relationship    race    sex  \
0   Married-civ-spouse     Exec-managerial         Husband   White   Male   
1             Divorced   Handlers-cleaners   Not-in-family   White   Male   
2   Married-civ-spouse   Handlers-cleaners         Husband   Black   Male   

   capital-gain  capital-loss  hours-per-week  native-country  income  
0             0             0              13   United-States   <=50K  
1             0             0              40   United-States   <=50K  
2             0             0              40   United-States   <=50K

In [15]:
"""
tail() es la análoga a head, pero mostrando las útlimas
"""
df.tail()

age      workclass  fnlwgt    education  education-num  \
32555   27        Private  257302   Assoc-acdm             12   
32556   40        Private  154374      HS-grad              9   
32557   58        Private  151910      HS-grad              9   
32558   22        Private  201490      HS-grad              9   
32559   52   Self-emp-inc  287927      HS-grad              9   

            marital-status          occupation relationship    race      sex  \
32555   Married-civ-spouse        Tech-support         Wife   White   Female   
32556   Married-civ-spouse   Machine-op-inspct      Husband   White     Male   
32557              Widowed        Adm-clerical    Unmarried   White   Female   
32558        Never-married        Adm-clerical    Own-child   White     Male   
32559   Married-civ-spouse     Exec-managerial         Wife   White   Female   

       capital-gain  capital-loss  hours-per-week  native-country  income  
32555             0             0              38   United-States   <=50K  
32556             0             0              40   United-States    >50K  
32557             0             0              40   United-States   <=50K  
32558             0             0              20   United-States   <=50K  
32559         15024             0              40   United-States    >50K

In [16]:
df.tail(n=2)

age      workclass  fnlwgt education  education-num  \
32558   22        Private  201490   HS-grad              9   
32559   52   Self-emp-inc  287927   HS-grad              9   

            marital-status        occupation relationship    race      sex  \
32558        Never-married      Adm-clerical    Own-child   White     Male   
32559   Married-civ-spouse   Exec-managerial         Wife   White   Female   

       capital-gain  capital-loss  hours-per-week  native-country  income  
32558             0             0              20   United-States   <=50K  
32559         15024             0              40   United-States    >50K

In [17]:
"""
Extraer muestra de filas aleatorias. df.sample()
Es al azar, cada vez que recargas la celda será una diferente, aunque se puedon modificar ciertos parámetros.
n=nº filas a mostrar
frac=0. porcentaje del total a mostrar. Este parámetro es incompatible con n
random_tate=nº fija un punto de partida a partir del cual siempre obtendremos los mismos resultados. Se trata de fijar una semilla. por lo que siempre obtendremos los mismos datos
                la seleccion ha sido aleatorio, pero permite replicar resultados.
                Este parámetro está disponible en otros muchos métodos.
"""
df.sample()

age workclass  fnlwgt   education  education-num       marital-status  \
6004   41   Private  182567   Bachelors             13   Married-civ-spouse   

            occupation relationship    race    sex  capital-gain  \
6004   Exec-managerial      Husband   White   Male             0   

      capital-loss  hours-per-week native-country income  
6004             0              40              ?   >50K

In [18]:
df.sample(4)

age     workclass  fnlwgt      education  education-num  \
9198    35       Private  374524    Prof-school             15   
29144   38       Private  234807      Bachelors             13   
6452    42   Federal-gov  170230        Masters             14   
17864   49     State-gov  118567   Some-college             10   

            marital-status        occupation    relationship    race      sex  \
9198    Married-civ-spouse    Prof-specialty         Husband   White     Male   
29144   Married-civ-spouse   Exec-managerial            Wife   White   Female   
6452         Never-married   Exec-managerial   Not-in-family   White   Female   
17864   Married-civ-spouse   Farming-fishing         Husband   White     Male   

       capital-gain  capital-loss  hours-per-week  native-country  income  
9198              0             0              75   United-States    >50K  
29144             0             0              50   United-States    >50K  
6452          14084             0              60   United-States    >50K  
17864             0             0              40   United-States   <=50K

In [21]:
df.sample(frac=0.01)

age     workclass  fnlwgt      education  education-num  \
1019    41       Private  198316   Some-college             10   
3511    26       Private  152035        HS-grad              9   
6367    22       Private  185279        HS-grad              9   
25305   56   Federal-gov  277420      Bachelors             13   
6430    34   Federal-gov  112115   Some-college             10   
...    ...           ...     ...            ...            ...   
12756   34       Private  236318        HS-grad              9   
7328    33       Private  132832        HS-grad              9   
22510   51       Private  173291        HS-grad              9   
29382   46       Private   65353   Some-college             10   
2251    44       Private  175935      Doctorate             16   

            marital-status          occupation     relationship    race  \
1019         Never-married        Craft-repair    Not-in-family   White   
3511         Never-married        Adm-clerical        Own-child   White   
6367         Never-married   Handlers-cleaners        Own-child   White   
25305   Married-civ-spouse     Exec-managerial          Husband   White   
6430         Never-married        Craft-repair   Other-relative   White   
...                    ...                 ...              ...     ...   
12756        Never-married   Machine-op-inspct        Unmarried   White   
7328         Never-married        Adm-clerical    Not-in-family   White   
22510             Divorced       Other-service    Not-in-family   White   
29382             Divorced    Transport-moving        Own-child   White   
2251    Married-civ-spouse        Adm-clerical          Husband   White   

           sex  capital-gain  capital-loss  hours-per-week  native-country  \
1019      Male             0             0              50           Japan   
3511    Female             0             0              40   United-States   
6367      Male             0             0              16   United-States   
25305     Male             0             0              40     Puerto-Rico   
6430      Male             0             0              40   United-States   
...        ...           ...           ...             ...             ...   
12756     Male             0             0              40   United-States   
7328    Female             0             0              50   United-States   
22510   Female             0             0              40   United-States   
29382     Male          3325             0              55   United-States   
2251      Male             0             0              55   United-States   

       income  
1019    <=50K  
3511    <=50K  
6367    <=50K  
25305    >50K  
6430    <=50K  
...       ...  
12756   <=50K  
7328    <=50K  
22510   <=50K  
29382   <=50K  
2251     >50K  

[326 rows x 15 columns]

In [22]:
df.sample(n=5, random_state=10)

age      workclass  fnlwgt      education  education-num  \
11358   41        Private  134130      Bachelors             13   
10859   38        Private  107302   Some-college             10   
30948   24        Private  136687        HS-grad              9   
29811   35   Self-emp-inc  187053      Bachelors             13   
18408   69              ?  254834      Bachelors             13   

            marital-status          occupation    relationship    race  \
11358   Married-civ-spouse               Sales         Husband   White   
10859   Married-civ-spouse               Sales         Husband   White   
30948            Separated   Machine-op-inspct       Unmarried   Other   
29811            Separated      Prof-specialty   Not-in-family   White   
18408   Married-civ-spouse                   ?         Husband   White   

           sex  capital-gain  capital-loss  hours-per-week  native-country  \
11358     Male             0             0              50   United-States   
10859     Male             0             0              60   United-States   
30948   Female             0             0              40   United-States   
29811   Female             0             0              50   United-States   
18408     Male         10605             0              10   United-States   

       income  
11358    >50K  
10859    >50K  
30948   <=50K  
29811   <=50K  
18408    >50K

In [23]:
"""
df.index -- no es un método, es un atributo.
Cuando no especificamos un índice en concreto, se genera un índice automático desde 0 hasta el último índice.
Aquí dice que se trata de un rango, que empieza en 0, acaba en 32560 con saltos de 1

dentro del propio índice, se puede obtener un array con los valores de los índices con atributo:
pd.index.values
"""

df.index

RangeIndex(start=0, stop=32560, step=1)

In [24]:
df.index.values

array([    0,     1,     2, ..., 32557, 32558, 32559], dtype=int64)

In [25]:
"""
Extracción de nombres de las columnas.
Mediante el atriburto del df columns
"""

df.columns

Index(['age', 'workclass', 'fnlwgt', 'education', 'education-num',
       'marital-status', 'occupation', 'relationship', 'race', 'sex',
       'capital-gain', 'capital-loss', 'hours-per-week', 'native-country',
       'income'],
      dtype='object')